src/Machine.Specifications.Runner.VisualStudio/MspecTestAdapterExecutor.cs

```diff
@@ -4,92 +4,73 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
-using System.Diagnostics;
-using System.IO;
 using Machine.VSTestAdapter.Execution;
 using Machine.VSTestAdapter.Helpers;
 using Machine.VSTestAdapter.Configuration;
 
 namespace Machine.VSTestAdapter
 {
-    public partial class MSpecTestAdapter : ITestExecutor
+    public class MSpecTestAdapterExecutor
     {
-        public void Cancel()
+        readonly ISpecificationExecutor executor;
+        readonly MSpecTestAdapterDiscoverer discover;
+        readonly ISpecificationFilterProvider specificationFilterProvider;
+
+        public MSpecTestAdapterExecutor(ISpecificationExecutor executor, MSpecTestAdapterDiscoverer discover, ISpecificationFilterProvider specificationFilterProvider)
         {
-            // Not supported
+            this.executor = executor;
+            this.discover = discover;
+            this.specificationFilterProvider = specificationFilterProvider;
         }
 
         public void RunTests(IEnumerable<string> sources, IRunContext runContext, IFrameworkHandle frameworkHandle)
         {
-            //Debugger.Launch();
-
-            frameworkHandle.SendMessage(TestMessageLevel.Informational, "Machine Specifications Visual Studio Test Adapter - Executing Specifications.");
-
-            Settings settings = GetSettings(runContext);
-
-            foreach (string currentAsssembly in sources.Distinct())
-            {
-                try
-                {
-#if !NETSTANDARD
-                    if (!File.Exists(Path.Combine(Path.GetDirectoryName(Path.GetFullPath(currentAsssembly)), "Machine.Specifications.dll")))
-                    {
-                        frameworkHandle.SendMessage(TestMessageLevel.Informational, String.Format("Machine.Specifications.dll not found for {0}", currentAsssembly));
-                        continue;
-                    }
-#endif
-
-                    frameworkHandle.SendMessage(TestMessageLevel.Informational, String.Format("Machine Specifications Visual Studio Test Adapter - Executing tests in {0}", currentAsssembly));
-
-                    this.executor.RunAssembly(currentAsssembly, settings, uri, frameworkHandle);
```

Comment:
This is the only change that makes me nervous, as it affects one of the hot paths for the MSpec runner. 99% of CI runners will use `RunAssembly` instead of the test-by-test runner, and I remember there were some oddities with the `DefaultRunner` and running behavior tests especially. I'll merge this, but only release it as beta for now until I can verify that nothing is broken. I've got 2000+ test solutions I can validate this against.

Reply:
I have been thinking about it over the weekend.
I would be happy to split the "new" and old paths so it worked 100% as before without a filter. I find it odd that the filtering is done in the Execution. (Why not in the Discovery?)

Can the "2000+ test solutions I can validate this against." be checked in? I was also going to add some filtering tests but it was hard to mock the Microsoft Filters.

I was also be tempted to add a timer (logged) to make sure finding tests isn't slow. NUnit appears to have lots of caching around it all (but could just be old?).

Reply:
Sadly not, the large 2000+ test solution is for work only.